### PR TITLE
fixes for bitmap update

### PR DESCRIPTION
**Pull request by jsorg71**
function not used for year,
part of noorders

### CODE REVIEW

File: libxrdp/libxrdp.c [hunk 1]
```diff
@@ -373,43 +375,59 @@ libxrdp_send_bell(struct xrdp_session *session)
...
+        case 16:
+            server_line_bytes = width * 2;
+            break;
+        case 24:
+        case 32:
+            server_line_bytes = width * 4;
+            break;
+        default: /* 8 bpp */
+            server_line_bytes = width;
+            break;
```
Comment (proski) [inline libxrdp/libxrdp.c]: Please define `server_Bpp` (any reasonable name) and use `server_line_bytes = width * server_Bpp`

File: libxrdp/libxrdp.c [hunk 2]
```diff
@@ -373,43 +375,59 @@ libxrdp_send_bell(struct xrdp_session *session)
...
     struct stream *temp_s = (struct stream *)NULL;
+    tui32 pixel;
 
-    DEBUG(("libxrdp_send_bitmap sending bitmap"));
+    LLOGLN(10, ("libxrdp_send_bitmap: sending bitmap"));
     Bpp = (bpp + 7) / 8;
-    e = width % 4;
-
-    if (e != 0)
+    e = (4 - width) & 3;
```
Comment (proski) [inline libxrdp/libxrdp.c]: Please consider a better name for that variable. `width_pad` maybe.

File: libxrdp/libxrdp.c [hunk 3]
```diff
@@ -421,6 +439,8 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
 
         while (i > 0)
         {
+             LLOGLN(10, ("libxrdp_send_bitmap: i %d", i));
```
Comment (proski) [inline libxrdp/libxrdp.c]: Indent looks a bit off here

File: libxrdp/libxrdp.c [hunk 4]
```diff
@@ -481,54 +518,73 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
                     out_uint16_le(s, j); /* line size */
                     j = j * lines_sending;
                     out_uint16_le(s, j); /* final size */
+                    total_bufsize += 26;
```
Comment (proski) [inline libxrdp/libxrdp.c]: Magic numbers 18 and 26

File: libxrdp/libxrdp.c [hunk 5]
```diff
@@ -481,54 +518,73 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
         p = data;
 
-        if (line_size > 0 && total_lines > 0)
+        if (line_bytes > 0 && total_lines > 0)
         {
             while (i < total_lines)
             {
-                lines_sending = 4096 / (line_size + e * Bpp);
+
+                lines_sending = (MAX_BITMAP_BUF_SIZE - 100) / (line_pad_bytes);
```
Comment (proski) [inline libxrdp/libxrdp.c]: Unneeded parentheses around line_pad_bytes

File: libxrdp/libxrdp.c [hunk 6]
```diff
@@ -481,54 +518,73 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
                 }
 
-                p = p + line_size * lines_sending;
+                if (lines_sending == 0)
+                {
+                    LLOGLN(0, ("libxrdp_send_bitmap: error, lines_sending == zero"));
+                    break;
+                }
+
+                p = p + server_line_bytes * lines_sending;
```
Comment (proski) [inline libxrdp/libxrdp.c]: `p +=` would look better to my taste.

File: libxrdp/libxrdp.c (hunk 5 above)
Comment (proski) [inline libxrdp/libxrdp.c]: Also, what is 100?

File: libxrdp/libxrdp.c (hunk 2 above)
Comment (jsorg71) [inline libxrdp/libxrdp.c]: I've been using e for "extra" but we can rename it later

File: libxrdp/libxrdp.c (hunk 1 above)
Comment (jsorg71) [inline libxrdp/libxrdp.c]: For this we can't just do a calculation because 24 and 32 are both 4 bytes / pixel.

File: libxrdp/libxrdp.c (hunk 4 above)
Comment (jsorg71) [inline libxrdp/libxrdp.c]: They are bytes since pop layer, I'll put comment

**Comment by metalefty** [comment]
Something's going wrong in Travis.  False CI failure.

File: libxrdp/libxrdp.c (hunk 1 above)
Comment (proski) [inline libxrdp/libxrdp.c]: I mean, you can define a new variable that would be 4 for 24 and 32.

File: libxrdp/libxrdp.c [hunk 7]
```diff
@@ -540,14 +596,67 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
+                            q = q - server_line_bytes;
+                            for (k = 0; k < width; k++)
+                            {
+                                pixel = *((tui16*)(q + k * 2));
+                                out_uint16_le(s, pixel);
+                            }
+                            for (k = 0; k < e; k++)
+                            {
+                                out_uint8s(s, 2);
+                            }
```
Comment (proski) [inline libxrdp/libxrdp.c]: Cannot you just use `out_uint8s(s, 2 * e)` here?

File: libxrdp/libxrdp.c [hunk 8]
```diff
@@ -540,14 +596,67 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
+                            {
+                                pixel = *((tui32*)(q + k * 4));
+                                out_uint8(s, pixel);
+                                out_uint8(s, pixel >> 8);
+                                out_uint8(s, pixel >> 16);
+                            }
+                            for (k = 0; k < e; k++)
+                            {
+                                out_uint8s(s, 3);
+                            }
```
Comment (proski) [inline libxrdp/libxrdp.c]: Likewise

File: libxrdp/libxrdp.c [hunk 9]
```diff
@@ -540,14 +596,67 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
+                            q = q - server_line_bytes;
+                            for (k = 0; k < width; k++)
+                            {
+                                pixel = *((int*)(q + k * 4));
+                                out_uint32_le(s, pixel);
+                            }
+                            for (k = 0; k < e; k++)
+                            {
+                                out_uint8s(s, 4);
+                            }
```
Comment (proski) [inline libxrdp/libxrdp.c]: Likewise

File: libxrdp/libxrdp.c [hunk 10]
```diff
@@ -540,14 +596,67 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
+                        }
+                        break;
+                    default: /* 8 bpp */
+                        for (j = 0; j < lines_sending; j++)
+                        {
+                            q = q - line_bytes;
+                            out_uint8a(s, q, line_bytes);
+                            out_uint8s(s, e * Bpp);
```
Comment (proski) [inline libxrdp/libxrdp.c]: We know Bpp is 1, we can omit it.

**Comment by metalefty** [comment]
I fixed travis at #552. Rebasing to current devel head will get build success.

File: libxrdp/libxrdp.c [hunk 11]
```diff
@@ -540,14 +596,58 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
+                            out_uint8s(s, e * 3);
+                        }
+                        break;
+                    case 32:
+                        for (j = 0; j < lines_sending; j++)
+                        {
+                            q = q - server_line_bytes;
+                            for (k = 0; k < width; k++)
+                            {
+                                pixel = *((int*)(q + k * 4));
```
Comment (proski) [inline libxrdp/libxrdp.c]: Spacing:
`pixel = *((int *) (q + k * 4));`

I'm not a fan of `tui` types (we can change to `uint32_t` later), but let's use either `int` or `tui32` consistently in this function

File: libxrdp/libxrdp.c [hunk 12]
```diff
@@ -540,14 +596,58 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
+                            out_uint8s(s, e * 2);
+                        }
+                        break;
+                    case 24:
+                        for (j = 0; j < lines_sending; j++)
+                        {
+                            q = q - server_line_bytes;
+                            for (k = 0; k < width; k++)
+                            {
+                                pixel = *((tui32*)(q + k * 4));
```
Comment (proski) [inline libxrdp/libxrdp.c]: Spacing:
`pixel = *((tui32 *) (q + k * 4));`

File: libxrdp/libxrdp.c [hunk 13]
```diff
@@ -540,14 +596,58 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
...
+                        }
+                        break;
+                    case 15:
+                    case 16:
+                        for (j = 0; j < lines_sending; j++)
+                        {
+                            q = q - server_line_bytes;
+                            for (k = 0; k < width; k++)
+                            {
+                                pixel = *((tui16*)(q + k * 2));
```
Comment (proski) [inline libxrdp/libxrdp.c]: Spacing:
`pixel = *((tui16 *) (q + k * 2));`

**Comment by proski** [comment]
It's more important that we have a day or two for testing than that we get the spacing right. We can always reindent to code after the release. Likewise, the bpp to Bpp conversion code can be done in two inline functions - after the release.

**Comment by metalefty** [comment]
Merging as is.

**Pull request merged**